File: akka-http-core/src/main/scala/akka/http/model/headers/HttpChallenge.scala
/**
 * Copyright (C) 2009-2014 Typesafe Inc. <http://www.typesafe.com>
 */

package akka.http.model
package headers

import akka.http.util._
import java.util

import akka.http.model.japi.JavaMapping.Implicits._

final case class HttpChallenge(scheme: String, realm: String,
                               params: Map[String, String] = Map.empty) extends japi.headers.HttpChallenge with ValueRenderable {

  def render[R <: Rendering](r: R): r.type = {
    r ~~ scheme ~~ " realm=" ~~# realm
    if (params.nonEmpty) params.foreach { case (k, v) ⇒ r ~~ ',' ~~ k ~~ '=' ~~# v }
    r
  }

  /** Java API */
  def getParams: util.Map[String, String] = params.asJava
}

File: akka-http-core/src/main/java/akka/http/model/japi/headers/HttpChallenge.java
/**
 * Copyright (C) 2009-2014 Typesafe Inc. <http://www.typesafe.com>
 */

package akka.http.model.japi.headers;

import akka.http.model.headers.HttpChallenge$;
import akka.http.model.japi.Util;

import java.util.Map;

public abstract class HttpChallenge {
    public abstract String scheme();
    public abstract String realm();

    public abstract Map<String, String> getParams();

    public static HttpChallenge create(String scheme, String realm) {
        return new akka.http.model.headers.HttpChallenge(scheme, realm, Util.emptyMap);
    }
    public static HttpChallenge create(String scheme, String realm, Map<String, String> params) {
        return new akka.http.model.headers.HttpChallenge(scheme, realm, Util.convertMapToScala(params));
    }
}